### PR TITLE
Bugfix/fix data loading

### PR DESCRIPTION
**Pull request by xantorres**
This should fix the issues around the new contextLoader/Updater helpers and specially with clusterContextLoader/Updater.

There are also improvements with these helpers, like now having `apiClient` and `loadFromContext` in `contextLoader` args
image

And `apiClient` and `currentItems` in `contextUpdater` args
image

### CODE REVIEW

File: src/app/core/helpers/clusterContextLoader.js [hunk 1]
```diff
@@ -1,24 +1,22 @@
-import { pathOr, chain } from 'ramda'
+import { pathOr, prop, groupBy } from 'ramda'
 import { loadClusters } from 'k8s/components/infrastructure/actions'
 import contextLoader from 'core/helpers/contextLoader'
 
 // Returns a contextLoaded function contextualized by selected cluster (given by clusterId param)
```
Comment (brennancheung) [inline src/app/core/helpers/clusterContextLoader.js]: This is a bit hard to follow.  Elaborating a little more and also a usage example in the comments would help clarify.

File: src/app/core/helpers/clusterContextLoader.js [hunk 2]
```diff
@@ -1,24 +1,22 @@
...
 const clusterContextLoader = (key, loaderFn, defaultValue = []) =>
-  async ({ context, params = {}, ...rest }) => {
-    const clusters = (await loadClusters({ context, params, ...rest, reload: false }))
+  async ({ context, setContext, params = {}, ...rest }) => {
+    const clusters = (await loadClusters({ context, setContext, params, ...rest, reload: false }))
       .filter(cluster => cluster.hasMasterNode)
```
Comment (brennancheung) [inline src/app/core/helpers/clusterContextLoader.js]: Hmm, I'm wondering if we always want to exclude clusters without master nodes.  Perhaps this should be an option and have it default to true.

File: src/app/core/helpers/contextLoader.js [hunk 3]
```diff
@@ -1,46 +1,53 @@
...
  * Returns a function that will use context to load and cache values
- * @param pathResolver Context path on which the resolved value will be cached, it can be a function
+ * @param contextPath Context path (a string or array for deep paths) on which the resolved value will be cached
  * @param loaderFn Function returning the data to be assigned to the context
  * @param defaultValue Default value assigned to the context
  * @returns {Function}
  */
-const contextLoader = (pathResolver, loaderFn, defaultValue = []) => {
-  const moizedPathResolver = moize(ensureFunction(pathResolver))
+const contextLoader = (contextPath, loaderFn, defaultValue = []) => {
```
Comment (brennancheung) [inline src/app/core/helpers/contextLoader.js]: Should any mapping—converting from the API version of an object, to the version needed inside the UI happen in the `loaderFn`?  This mapping function needs to be applied when loading, and also when create/update calls return an object.

File: src/app/core/helpers/contextLoader.js [hunk 4]
```diff
@@ -1,46 +1,53 @@
...
-import { ensureFunction, ensureArray } from 'utils/fp'
+import { ensureArray } from 'utils/fp'
 
 let pendingPromises = {}
+let resolvers = {}
 
 /**
  * Returns a function that will use context to load and cache values
- * @param pathResolver Context path on which the resolved value will be cached, it can be a function
+ * @param contextPath Context path (a string or array for deep paths) on which the resolved value will be cached
```
Comment (brennancheung) [inline src/app/core/helpers/contextLoader.js]: Good call.  This adds a lot of flexibility.

File: src/app/core/helpers/contextLoader.js [hunk 5]
```diff
@@ -1,46 +1,53 @@
...
-      dissocPath(resolvedPath, pendingPromises)
+      await setContext(prevContext => assocPath(arrContextPath, output, prevContext))
+      pendingPromises = dissocPath(arrContextPath, pendingPromises)
     }
     return output
   }
+
+  resolvers = assocPath(arrContextPath, resolver, resolvers)
+
+  return resolver
```
Comment (brennancheung) [inline src/app/core/helpers/contextLoader.js]: Add comment on what uses this return value and why it needs it.

File: src/app/plugins/kubernetes/components/apps/actions.js [hunk 6]
```diff
@@ -1,31 +1,24 @@
...
 
 export const loadApps = clusterContextLoader(
   'apps',
   async ({ context, clusters, params: { clusterId } }) => {
-    const loadClusterAppsFromContext = getClusterApps(context)
+    const { qbert } = context.apiClient
     return clusterId === '__all__'
-      ? Promise.all(clusters.map(loadClusterAppsFromContext)).then(flatten)
-      : loadClusterAppsFromContext(clusters.find(propEq('uuid', clusterId)))
+      ? asyncFlatMap(pluck('uuid', clusters), qbert.getCharts)
```
Comment (brennancheung) [inline src/app/plugins/kubernetes/components/apps/actions.js]: Very minor and optional, do we want to adopt the convention we have used elsewhere of `{ service: 'qbert', entity: 'charts' }`?

File: src/app/core/helpers/contextLoader.js (hunk 3 above)
Comment (xantorres) [reply src/app/core/helpers/contextLoader.js]: For this case I would suggest having a third argument in the `contextLoader` function, this argument would be a mapper that will receive the loaded data and will do any extra transformations, also this same function would be applied when using `contextUpdater`.

Comment (brennancheung) [reply src/app/core/helpers/contextLoader.js]: Yes, but I think we might want to refactor this to be an object `fn({ contextPath, loaderFn, ... })` instead of individual positional parameters.  Having 4 function parameters is too many when it is positional.

Comment (xantorres) [reply src/app/core/helpers/contextLoader.js]: I agree but still we will have to remember that this object always requires the `contextPath` and `loaderFn` props, what do you think about leaving the first two arguments as function parameters and have a third `options` object parameter with these additional optional things?

File: src/app/plugins/kubernetes/components/apps/actions.js (hunk 6 above)
Comment (xantorres) [reply src/app/plugins/kubernetes/components/apps/actions.js]: Good point, I will look into it in another PR
